src/testutils/helpers.rs

```diff
@@ -0,0 +1,200 @@
+// Bitcoin Dev Kit
+// Written in 2020 by Alekos Filini <alekos.filini@gmail.com>
+//
+// Copyright (c) 2020-2021 Bitcoin Dev Kit Developers
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+#![allow(missing_docs)]
+#![allow(unused)]
+
+use std::str::FromStr;
+
+use bitcoin::{Address, Network, OutPoint, Transaction, TxIn, TxOut, Txid};
+
+use crate::{
+    database::{AnyDatabase, BatchDatabase, BatchOperations, MemoryDatabase},
+    testutils, BlockTime, KeychainKind, LocalUtxo, TransactionDetails, Wallet,
+};
+
+use super::TestIncomingTx;
+
+/// Populate a test database with a `TestIncomingTx`, as if we had found the tx with a `sync`.
+/// This is a hidden function, only useful for `DataBase` unit testing.
+pub fn populate_test_db(
+    db: &mut impl BatchOperations,
+    tx_meta: TestIncomingTx,
+    current_height: u32,
+    is_coinbase: bool,
+) -> Txid {
+    // Ignore `tx_meta` inputs while creating a coinbase transaction
+    let input = if is_coinbase {
+        // `TxIn::default()` creates a coinbase input, by definition.
+        vec![TxIn::default()]
+    } else {
+        tx_meta
+            .input
+            .iter()
+            .map(|test_input| {
+                let mut txin = TxIn {
+                    previous_output: OutPoint {
+                        txid: test_input.txid,
+                        vout: test_input.vout,
+                    },
+                    ..Default::default()
+                };
+
+                if let Some(seq) = test_input.sequence {
+                    txin.sequence = seq;
+                }
+                txin
+            })
+            .collect()
+    };
+
+    let output = tx_meta
+        .output
+        .iter()
+        .map(|out_meta| TxOut {
+            value: out_meta.value,
+            script_pubkey: Address::from_str(&out_meta.to_address)
+                .unwrap()
+                .script_pubkey(),
+        })
+        .collect();
+
+    let tx = Transaction {
+        version: 1,
+        lock_time: 0,
+        input,
+        output,
+    };
+
+    let txid = tx.txid();
+
+    let confirmation_time = if let Some(conf_time) = tx_meta.min_confirmations {
+        current_height
+            .checked_sub(conf_time as u32)
+            .map(|conf| BlockTime {
+                height: conf,
+                timestamp: 0,
+            })
+    } else {
+        None
+    };
```

Comment:
I think this is very different from what we had before:

- In the macro if `min_confirmations` is zero we map that to `None` immediately
- The macro had an `unwrap()` on the `checked_sub()`. I understand it's better to handle errors, but in this case we want to fail if the height is too low for a given number of confirmations. This code instead silently makes the tx unconfirmed because it maps that case to `None`
- There's a `+1` missing which was present in the macro

Reply:
> In the macro if min_confirmations is zero we map that to None immediately


Is that the correct behavior though?? If minimum confirmation is 0, and current height is 100, I would expect tx_meta confirmation height to be at 100. We were setting it to `None` previously even when we were setting a transaction with min conf 0, which should make the transaction immediately confirmed. Maybe I am confused on the intention of min_confirmation here.. 

> The macro had an unwrap() on the checked_sub(). I understand it's better to handle errors, but in this case we want to fail if the height is too low for a given number of confirmations. This code instead silently makes the tx unconfirmed because it maps that case to None


In my head this was actually the intended behavior I thought of.. There is no error situation here.. If we set a transaction with min_conf < current height, it simply means the transaction isn't confirmed yet. This is used in `test_bump_fee_absolute_force_add_input` test to make an unconfirmed transaction. That was previously using the min_conf=0 value, which was confusing to me. And I ended up changing the logic while trying to understand whats going on.. If a transaction with `min_conf = 1` makes it confirmed at height `current_height - 1` I would expect `min_conf=0` would make it confirmed at `current_height`. Not an unconfirmed transaction.

> There's a +1 missing which was present in the macro


May be I am missing something.. But why is this +1 required? If current height is 100, min_conf is 5. The tx confirmation height would be at 95. Which is simply the subtraction of two numbers.

Reply:
> May be I am missing something.. But why is this +1 required? If current height is 100, min_conf is 5. The tx confirmation height would be at 95. Which is simply the subtraction of two numbers.

Nope: if the current height is 100, and my transaction was included in block 100, then it has 1 confirmation. If the current height is 100, and my transaction is still in mempool, then it has 0 confirmations.

> Is that the correct behavior though?? If minimum confirmation is 0, and current height is 100, I would expect tx_meta confirmation height to be at 100. 

Nope, if the confirmation is 0 then the transaction was never included in a block, so there's no confirmation_time

>  If a transaction with min_conf = 1 makes it confirmed at height current_height - 1 I would expect min_conf=0 would make it confirmed at current_height. Not an unconfirmed transaction.

Well, but min_conf = 1 makes it confirmed at height `current_height`

Reply:
Hey @danielabrozzoni I updated my comments before I saw yours.. Yes it makes sense and I am trying to revert back to the previous behavior.. My bad.. I missed the +1 calculation there..